cpp/src/parquet/schema.cc

```diff
@@ -795,12 +795,27 @@ void SchemaDescriptor::Init(NodePtr schema) {
 }
 
 bool SchemaDescriptor::Equals(const SchemaDescriptor& other) const {
+  return this->Equals(other, nullptr);
+}
+
+bool SchemaDescriptor::Equals(
+    const SchemaDescriptor& other,
+    std::shared_ptr<std::stringstream> diff_msg = NULLPTR) const {
   if (this->num_columns() != other.num_columns()) {
+    if (diff_msg != nullptr) {
+      *diff_msg.get() << "This schema has " << this->num_columns()
+                      << " columns, other has " << other.num_columns();
+    }
     return false;
   }
 
   for (int i = 0; i < this->num_columns(); ++i) {
     if (!this->Column(i)->Equals(*other.Column(i))) {
+      if (diff_msg != nullptr) {
+        *diff_msg.get() << "These two columns differ:" << std::endl
+                        << this->Column(i)->ToString() << std::endl
+                        << other.Column(i)->ToString();
```

Comment:
Sorry if this is a silly question, but, `ToString() `uses `PrettyPrint` to print out the column values. It prints out the schema also, right? Can you maybe share the message from Python when two columns differ?

Reply:
Good point, I should have presented a short example. Please see my updated comment for this PR. 👍

Reply:
Do you think it would make any sense to replace "This schema ..." with something more informative? I think in the test, `this` is `meta1` and `other` is `meta2`? I think that would help when debugging the error.

Similar for "These" being `meta1` and `meta2` columns on position `i-1` differ?

Reply:
Sure thing. Were you thinking something like _"The first schema has N columns, the second has J columns"_ then? For me, "this" would refer to the one being appended to, but likely I have tunnel vision in this case. ;-) 

For the "These columns differ".. you're thinking adding the index  of the two then? Do you think the resulting printed schemas of the two columns are sufficient?

Reply:
I hope I didn't complicate things for you!

In the example you added I was thinking something like:
```python
# meta1 and meta2 differ in number of columns
meta1.append_row_groups(meta2)
*** RuntimeError: AppendRowGroups requires equal schemas but meta1 has 2 columns and meta2 has 1.
```

```python
# meta1 and meta2 differ in column types
meta1.append_row_groups(meta2)
*** RuntimeError: AppendRowGroups requires equal schemas but schemas meta1 and meta2 differ in these columns:
...
```

But note, this is just me thinking about it. Feel free to ignore it ;)

Reply:
ah, I'm following now. I'm not sure we have access to the specific variable names inside of this C++ context, but I'll see what's possible. 👍

Reply:
I don't think there is an obvious way to get the exact variable names which caused the exception. We'd at least need to catch the exception raised from C++ then introspect somehow in Python. I'll resolve this(?) even.. but of course open to suggestions, I just couldn't make it work. 🤔